ui/shared/url-util.js

```diff
@@ -0,0 +1,177 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Modified from the following fils:
+//  * http://mxr.mozilla.org/gaia/source/shared/js/url_helper.js
+//  * https://github.com/brave/browser-laptop/blob/master/js/lib/urlutil.js
+// Changed to use require('url').parse instead of window.URL and document.createElement('a')
+
+// characters, then : with optional //
+const rscheme = /^(?:[a-z\u00a1-\uffff0-9-+]+)(?::(\/\/)?)/i;
+const defaultScheme = 'http://';
+const fileScheme = 'file://';
+const os = require('os');
+const urlParse = require('url').parse;
+
+/**
+ * A simple class for parsing and dealing with URLs.
+ * @class UrlUtil
+ */
+const UrlUtil = {
+
+  /**
+   * Extracts the scheme from a value.
+   * @param {String} input The input value.
+   * @returns {String} The found scheme.
+   */
+  getScheme(input) {
+    // This function returns one of following:
+    // - scheme + ':' (ex. http:)
+    // - scheme + '://' (ex. http://)
+    // - null
+    const scheme = (rscheme.exec(input) || [null])[0];
+    return scheme === 'localhost://' ? null : scheme;
+  },
+
+  /**
+   * Checks if an input has a scheme (e.g., http:// or ftp://).
+   * @param {String} input The input value.
+   * @returns {Boolean} Whether or not the input has a scheme.
+   */
+  hasScheme(input) {
+    return !!this.getScheme(input);
+  },
+
+  /**
+   * Prepends file scheme for file paths, otherwise the default scheme
+   * @param {String} input path, with opetional schema
+   * @returns {String} path with a scheme
+   */
+  prependScheme(input) {
+    // expand relative path
+    if (input.startsWith('~/')) {
+      input = input.replace(/^~/, os.homedir());
```

Comment:
How does `os.homedir()` work when this runs in content process? Or in the main process for that matter? These are all going to be transpiled during build and not processed at runtime, leaving most `os` commands useless or misleading. To detect current os/system/arch, we should use `build-config.js` (see current `shared/util.js` using `development` or `packaged`)